File: app/src/main/java/ru/nsu/bobrofon/easysshfs/log/LogFragment.java
package ru.nsu.bobrofon.easysshfs.log;


import android.app.Activity;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import ru.nsu.bobrofon.easysshfs.DrawerStatus;
import ru.nsu.bobrofon.easysshfs.EasySSHFSActivity;
import ru.nsu.bobrofon.easysshfs.R;

public class LogFragment extends Fragment implements LogModel.Observer {
	private static final String TAG = "LogFragment";

	private TextView mLogTextView;
	private LogModel mLogModel;
	private DrawerStatus mDrawerStatus;

	public LogFragment() {
		// Required empty public constructor
	}

	public void setDrawerStatus(final DrawerStatus drawerStatus) {
		mDrawerStatus = drawerStatus;
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		Log.i(TAG, "onCreateView");
		super.onCreateView(inflater, container, savedInstanceState);
		setHasOptionsMenu(true);

		// Inflate the layout for this fragment
		View selfView = inflater.inflate(R.layout.fragment_log, container, false);

		mLogTextView = selfView.findViewById(R.id.log);

		mLogModel = LogSingleton.getLogModel();

		mLogModel.registerObserver(this);

		return selfView;
	}

	@Override
	public void onDestroyView () {
		Log.i(TAG, "onDestroyView");
		super.onDestroyView();
		mLogModel.unregisterObserver(this);
	}


	@Override
	public void onLogChanged(LogModel logModel) {
		final String logHeader = getString(R.string.debug_log_header);
		final String logBody = logModel.getLog();
		mLogTextView.setText(String.format("%s%s", logHeader, logBody));
	}

	@Override
	public void onAttach(Activity activity) {
		super.onAttach(activity);
		((EasySSHFSActivity) activity).onSectionAttached(R.string.debug_log_title);
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
		if (mDrawerStatus == null || !mDrawerStatus.isDrawerOpen()) {
			inflater.inflate(R.menu.log, menu);
		}
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		int id = item.getItemId();

		if (id == R.id.action_clean) {
			mLogModel.clean();
			return true;
		}

		return super.onOptionsItemSelected(item);
	}
}

File: app/src/main/java/ru/nsu/bobrofon/easysshfs/log/LogFragment.kt
// SPDX-License-Identifier: MIT
package ru.nsu.bobrofon.easysshfs.log

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import ru.nsu.bobrofon.easysshfs.EasySSHFSFragment
import ru.nsu.bobrofon.easysshfs.R
import ru.nsu.bobrofon.easysshfs.databinding.FragmentLogBinding

private const val TAG = "LogFragment"

class LogFragment : EasySSHFSFragment(), LogChangeObserver {

    private val appLog = AppLog.instance()

    private var _binding: FragmentLogBinding? = null
    private val binding get() = _binding!!
    private val logTextView get() = binding.log


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        Log.i(TAG, "onCreateView")
        super.onCreateView(inflater, container, savedInstanceState)
        _binding = FragmentLogBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        appLog.registerObserver(this)

        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(menuProvider, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    override fun onDestroyView() {
        Log.i(TAG, "onDestroyView")
        appLog.unregisterObserver(this)
        super.onDestroyView()
        _binding = null
    }


    override fun onLogChanged(logView: LogView) {
        val logHeader = getString(R.string.debug_log_header)
        val logBody = logView.toString()
        logTextView.text = String.format("%s%s", logHeader, logBody)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        appActivity?.onSectionAttached(R.string.debug_log_title)
    }

    private val menuProvider = object : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            if (!drawerStatus.isDrawerOpen) {
                menuInflater.inflate(R.menu.log, menu)
            }
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
            val id = menuItem.itemId

            if (id == R.id.action_clean) {
                appLog.clean()
                return true
            }

            return false
        }
    }
}// Required empty public constructor
